mixnet/node/src/lib.rs

```diff
@@ -244,70 +241,96 @@ impl PacketForwarder {
+    }
+}
+
+struct Forwarder {
+    tx: mpsc::UnboundedSender<Body>,
+}
+
+impl Forwarder {
+    fn new(addr: SocketAddr, config: MixnetNodeConfig) -> Self {
+        let (tx, rx) = mpsc::unbounded_channel();
+
+        let tx_ = tx.clone();
+        tokio::spawn(async move {
+            Forwarder::run(addr, rx, tx_, config).await;
+        });
+
+        Self { tx }
+    }
+
+    async fn run(
+        addr: SocketAddr,
+        mut rx: mpsc::UnboundedReceiver<Body>,
+        mut tx: mpsc::UnboundedSender<Body>,
+        config: MixnetNodeConfig,
+    ) {
+        loop {
+            match Forwarder::connect_and_forward(addr, &mut rx, &mut tx).await {
+                Ok(_) => {
+                    tracing::debug!("closing the forwarder: {addr}");
+                    return;
                 }
                 Err(e) => {
-                    tracing::error!("failed to connect to {}: {e}", target);
-                    return Err(MixnetNodeError::Protocol(e.into()));
+                    tracing::error!("retrying: failed to connect and forward to {addr}: {e}");
+                    tokio::time::sleep(config.retry_delay).await;
                 }
             }
         }
-        Ok(body
-            .write(self.connections.get_mut(&target).unwrap())
-            .await?)
     }
 
-    async fn send(&mut self, pkt: Packet) {
-        if let Err(err) = self.try_send(pkt.target, &pkt.body).await {
-            match err {
-                MixnetNodeError::Protocol(ProtocolError::IO(e))
-                    if e.kind() == std::io::ErrorKind::Unsupported =>
-                {
-                    tracing::error!("fail to send message to {}: {e}", pkt.target);
+    async fn connect_and_forward(
+        addr: SocketAddr,
+        rx: &mut mpsc::UnboundedReceiver<Body>,
+        tx: &mut mpsc::UnboundedSender<Body>,
+    ) -> core::result::Result<(), ProtocolError> {
+        match TcpStream::connect(addr).await {
+            Ok(mut conn) => {
+                while let Some(body) = rx.recv().await {
+                    if let Err(e) = body.write(&mut conn).await {
+                        match e {
+                            ProtocolError::IO(_) => {
+                                tx.send(body).expect("the receiver half is always open");
+                                return Err(e);
+                            }
+                            _ => {
+                                tracing::error!("ignoring: failed to forward body to {addr}: {e}",);
+                            }
+                        }
+                    } else {
+                        tracing::debug!("forwarded body to {addr}");
+                    }
                 }
-                _ => self.handle_retry(pkt),
+
+                // the sender half has been closed.
+                Ok(())
             }
+            Err(e) => Err(ProtocolError::from(e)),
         }
     }
 
-    fn handle_retry(&self, mut pkt: Packet) {
```

Comment:
why did we remove retries?

Reply:
Ah, I forgot to mention about this in the PR. Actually, we're not removing retry. I changed the retry mechanism little bit in this case.

Previously, we keep a `retry_count` for each packet, and if it still fails after `max_retries`, we drop the packet.  

But, in this PR, I changed the retry mechanism from per-packet to per-connection, since a mixnode must always maintain connections with the next mix layer. If we fail to send a packet to a certain connection, we push the packet back to the queue and try to re-establish the connection. So, it's reconnect precisely (rather than retry).
(Actually, our previous retry implementation had something wrong. It didn't re-establish the connection when the write call is failed with IO error. Then, the retry won't help.)

Also, I removed the `max_retries` parameter in this PR because the mixnode has to try its best to re-establish connections with the next layer. This will make the message queue full if the retry keeps failing, but I think it's the common case that can happen in other parts that use queues. Maybe, we can revive the `max_retries` parameter, and drop/block all messages for the dead connection if the reconnect keeps failing.

p.s.
I found that I removed the exponential backoff. I'll revive it.

Reply:
Makes sense, after all packets in a connection will likely suffer the same fate